File: ensmic/evaluation/figure_results.R
# Import libraries
library("data.table")
library("ggplot2")
library("tidyr")
library("grid")
library("gridExtra")
library("stringr")

# Configurations
path_results <- "/home/mudomini/projects/ensmic/results/"
datasets <- c("chmnist", "covid", "isic", "drd")
phases <- c("baseline", "augmenting", "stacking", "bagging")

path_eval <- file.path(path_results, "evaluation")
dir.create(path_eval, showWarnings=FALSE)

# Iterate over all phases
for (i in seq(1,4)){
  # Load best score data
  path_file <- file.path(path_results, "eval_tmp", paste(phases[i], ".", "all.csv", sep="", collapse=NULL))
  dt_score <- fread(path_file)
  # Preprocess
  dt_score$dataset <- toupper(dt_score$dataset)
  dt_score$dataset <- paste("Dataset:", dt_score$dataset)
  dt_score <- transform(dt_score, dataset=factor(dataset, levels=c("Dataset: CHMNIST","Dataset: COVID",
                                                                   "Dataset: ISIC","Dataset: DRD")))

  # Plot figure A
  plot_a <- ggplot(dt_score[metric=="F1"], aes(method, value, fill=dataset)) +
    geom_bar(stat="identity", position="dodge", col="black", width=0.7) +
    scale_y_continuous(breaks=seq(0, 1, 0.1), limits=c(0, 1)) +
    facet_wrap(. ~ dataset) + 
    coord_flip() +
    theme_bw() +
    theme(legend.position = "none") +
    xlab("") +
    ylab("Metric - F1")
  # png(file.path(path_eval, paste("figure.results.", phases[i], ".A.png", sep="")), width=1200, height=1000, res=180)
  # print(plot_a)
  # dev.off()
  
  # Load roc data
  path_file <- file.path(path_results, "eval_tmp", paste(phases[i], ".", "roc.csv", sep="", collapse=NULL))
  dt_roc <- fread(path_file)
  # Preprocess
  dt_roc$dataset <- toupper(dt_roc$dataset)
  dt_roc$dataset <- paste("Dataset: ", dt_roc$dataset, " - Best Method: ", dt_roc$method, sep="")
  dt_roc <- transform(dt_roc, dataset=factor(dataset, levels=unique(dt_roc$dataset)))
  
  # Plot figure B  
  dt_roc_split <- split(dt_roc, f=dt_roc$dataset)
  p1 <- ggplot(dt_roc_split[[1]], aes(FPR, TPR, color=class)) + 
    geom_line() +
    geom_abline(intercept=0, slope=1, color="black",
                linetype="dashed") +
    facet_wrap(~ dataset, nrow=1, ncol=1) +
    scale_x_continuous(breaks=seq(0, 1, 0.1), limits=c(0, 1)) +
    scale_y_continuous(breaks=seq(0, 1, 0.1), limits=c(0, 1)) +
    scale_color_discrete(name="Classes") +
    theme_bw() +
    xlab("False Positive Rate") +
    ylab("True Positive Rate")
  
  p2 <- p1 %+% dt_roc_split[[2]]
  p3 <- p1 %+% dt_roc_split[[3]]
  p4 <- p1 %+% dt_roc_split[[4]]
  
  g1 = ggplotGrob(p1)
  g2 = ggplotGrob(p2)
  g3 = ggplotGrob(p3)
  g4 = ggplotGrob(p4)
  g1$widths <- g4$widths
  g2$widths <- g4$widths
  g3$widths <- g4$widths
  
  # png(file.path(path_eval, paste("figure.results.", phases[i], ".B.png", sep="")), width=1250, height=800, res=120)
  # plot_b <- arrangeGrob(g1,g2,g3,g4)
  # grid.arrange(plot_b)
  # dev.off()
  
  # Plot figure A and B together
  title <- paste("Results for Ensemble Learning Technique: ", str_to_title(phases[i]), sep="")
  png(file.path(path_eval, paste("figure.results.", phases[i], ".big.png", sep="")), width=2000, height=800, res=110)
  layout <- rbind(c(1,1,1,2,2,3,3),
                  c(1,1,1,4,4,5,5))
  grid.arrange(plot_a,g1,g2,g3,g4, layout_matrix=layout, top=textGrob(title, gp=gpar(fontsize=20,font=1)))
  dev.off()
}

File: ensmic/evaluation/figure_results.py
#==============================================================================#
#  Author:       Dominik Müller                                                #
#  Copyright:    2021 IT-Infrastructure for Translational Medical Research,    #
#                University of Augsburg                                        #
#                                                                              #
#  This program is free software: you can redistribute it and/or modify        #
#  it under the terms of the GNU General Public License as published by        #
#  the Free Software Foundation, either version 3 of the License, or           #
#  (at your option) any later version.                                         #
#                                                                              #
#  This program is distributed in the hope that it will be useful,             #
#  but WITHOUT ANY WARRANTY; without even the implied warranty of              #
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the               #
#  GNU General Public License for more details.                                #
#                                                                              #
#  You should have received a copy of the GNU General Public License           #
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.       #
#==============================================================================#
#-----------------------------------------------------#
#                   Library imports                   #
#-----------------------------------------------------#
# External libraries
import os
import pandas as pd
import json
from ast import literal_eval
import numpy as np
from plotnine import *
# Internal libraries/scripts
from ensmic.data_loading import IO_Inference
from ensmic.utils.metrics import compute_metrics, compute_rawCM

#-----------------------------------------------------#
#                    Configurations                   #
#-----------------------------------------------------#
path_data = "/home/mudomini/projects/ensmic/data"
path_results = "/home/mudomini/projects/ensmic/results"
phases = ["baseline", "augmenting", "stacking", "bagging"]
datasets = ["chmnist", "covid", "isic", "drd"]

#-----------------------------------------------------#
#                     Gather Data                     #
#-----------------------------------------------------#
# Iterate over phases
for phase in phases:
    result_set = []
    result_method = []
    result_all = []
    # Iterate over each dataset
    for ds in datasets:
        # Load ground truth dictionary
        path_gt = os.path.join(path_data, ds + ".test.json")
        with open(path_gt, "r") as json_reader:
            gt_map = json.load(json_reader)

        # Get current path
        path_current = os.path.join(path_results, "phase_" + phase + "." + ds)
        # Identify best method
        if phase != "bagging":
            data = pd.read_csv(os.path.join(path_current, "evaluation",
                                            "results.test.averaged.csv"))
            data_std = pd.read_csv(os.path.join(path_current, "evaluation",
                                                "results.test.std.csv"))
            if phase == "stacking":
                data.rename(columns={"ensembler": "method"}, inplace=True)
                data_std.rename(columns={"ensembler": "method"}, inplace=True)
            else:
                data.rename(columns={"architecture": "method"}, inplace=True)
                data_std.rename(columns={"architecture": "method"}, inplace=True)
            data_std.rename(columns={"value": "std"}, inplace=True)
            data = pd.merge(data, data_std, on=["method", "metric"])
            data["std"] = np.where(data["std"] >= data["value"],
                                   data["value"],
                                   data["std"])
            data_f1 = data[data["metric"] == "F1"]
            best_dt = data_f1.iloc[data_f1["value"].argmax()]
            best_method = best_dt["method"]
        # For bagging: Identify best architecture & best method
        else:
            # iterate over all architectures
            best_score = 0
            for walk in os.listdir(path_current):
                if os.path.isfile(os.path.join(path_current, walk)) : continue
                # Load results
                data = pd.read_csv(os.path.join(path_current, walk, "evaluation",
                                                "results.test.averaged.csv"))
                data_std = pd.read_csv(os.path.join(path_current, walk, "evaluation",
                                                    "results.test.std.csv"))
                # Obtain best
                data.rename(columns={"ensembler": "method"}, inplace=True)
                data_std.rename(columns={"ensembler": "method"}, inplace=True)
                data_std.rename(columns={"value": "std"}, inplace=True)
                data = pd.merge(data, data_std, on=["method", "metric"])
                data["std"] = np.where(data["std"] >= data["value"],
                                       data["value"],
                                       data["std"])
                data_f1 = data[data["metric"] == "F1"]
                curr_dt = data_f1.iloc[data_f1["value"].argmax()]
                score = curr_dt["value"]
                if best_score > score : continue
                else:
                    best_score = score
                    best_architecture = walk
                    best_method = curr_dt["method"]
            # update path_current with best architecture
            path_current = os.path.join(path_current, best_architecture)

        # Create an Inference IO Interface
        if phase != "bagging" : path_inf = os.path.join(path_current,
                                            best_method, "inference.test.json")
        else : path_inf = os.path.join(path_current, "inference",
                            "inference." + best_method + ".pred.json")
        infIO = IO_Inference(None, path=path_inf)
        # Load predictions for samples
        inference = infIO.load_inference()
        # Load class names
        class_list = infIO.load_inference(index="legend")

        # Initialize lists for predictions and ground truth
        id = []
        gt = []
        pd_class = []
        pd_prob = []
        # Iterate over all samples of the testing set
        sample_list = inference.keys()
        for sample in sample_list:
            # Obtain ground truth and predictions
            id.append(sample)
            gt.append(np.argmax(gt_map[sample]))
            prediction = inference[sample]
            pd_class.append(np.argmax(prediction))
            pd_prob.append(prediction)

        #----------------------------#
        #        ROC Analysis        #
        #----------------------------#
        metrics = compute_metrics(gt, pd_class, pd_prob, class_list)
        # Parse metrics to Pandas dataframe
        metrics_tidy = pd.DataFrame.from_dict(metrics)
        metrics_tidy = metrics_tidy[["ROC_FPR", "ROC_TPR"]]
        metrics_tidy = metrics_tidy.transpose()
        metrics_tidy.columns = class_list

        # Cache dataframe
        result_set.append(metrics_tidy)
        result_method.append(best_method)
        data["dataset"] = ds
        result_all.append(data)

    cols = ["dataset", "method", "class", "FPR", "TPR"]
    df_results = pd.DataFrame(data=[], dtype=np.float64, columns=cols)
    # Iterate over dataset
    for i in range(0, len(datasets)):
        # Preprocess data into correct format
        roc_df = result_set[i].copy()
        roc_df = roc_df.transpose()
        roc_df = roc_df[["ROC_FPR", "ROC_TPR"]]
        roc_df = roc_df.apply(pd.Series.explode)
        # Append to result dataframe
        roc_df = roc_df.reset_index()
        roc_df.rename(columns={"index":"class",
                               "ROC_FPR":"FPR",
                               "ROC_TPR":"TPR"},
                      inplace=True)
        roc_df["method"] = result_method[i]
        roc_df["dataset"] = datasets[i]
        # Reorder columns
        roc_df = roc_df[cols]
        # Convert from object to float
        roc_df["FPR"] = roc_df["FPR"].astype(float)
        roc_df["TPR"] = roc_df["TPR"].astype(float)
        # Merge to global result dataframe
        df_results = df_results.append(roc_df, ignore_index=True)
    path_res_tmp = os.path.join(path_results, "eval_tmp")
    if not os.path.exists(path_res_tmp) : os.mkdir(path_res_tmp)
    # Store ROC data
    path_res = os.path.join(path_res_tmp, phase + ".roc.csv")
    df_results.to_csv(path_res, index=False)
    # Store all data
    path_res_all = os.path.join(path_res_tmp, phase + ".all.csv")
    pd.concat(result_all).to_csv(path_res_all, index=False)
